kernel/arch/dreamcast/kernel/rtc.c

```diff
@@ -52,23 +63,66 @@ time_t rtc_unix_secs() {
     }
 
     /* Subtract out 20 years */
-    rtcnew = rtcnew - 631152000;
+    rtcnew = rtcnew - RTC_UNIX_EPOCH_DELTA;
 
     return rtcnew;
 }
 
+/* Sets the date/time value from a UNIX epoch time stamp, 
+   returning 0 for success or -1 for failure. */
+int rtc_set_unix_secs(time_t secs) {
+    int result = 0;
+    uint32_t rtcnew;
+    int i;
+
+    /* Adjust by 20 years to get to the expected RTC time. */
+    const uint32_t adjusted = secs + RTC_UNIX_EPOCH_DELTA;
+
+    /* Enable writing by setting LSB of control */
+    g2_write_32(RTC_CONTROL_ADDR, 0x1);
+
+    /* Try 3 times to ensure we didn't write a value then have 
+       the clock increment itself before the next. */
+    for(i = 0; i < RTC_RETRY_COUNT; i++) { 
+        /* Write the least-significant 16-bits first, because 
+           writing to the high 16-bits will lock RTC writes. */
+        g2_write_32(RTC_SECONDS_LOW_ADDR, (adjusted) & 0xffff);
+        g2_write_32(RTC_SECONDS_HIGH_ADDR, (adjusted >> 16) & 0xffff);
+
+        /* Read the time back again, to ensure it was written properly. */
```

Comment:
Should this just call rtc_unix_secs? If the reading is fragile, then couldn't it be possible that the write succeeds properly, but the read to verify it is broken without doing the looping? Then we loop over and do 2 invalid writes and maybe still not validate?

If the writing is fragile, as the comments suggest, I could also see a situation where on the last retry the write succeeds then we only get one chance to read correctly.